### ISSUE TITLE
Race conditions in Accumulator/models code

### ISSUE THREAD

**Issue by prydin** [influxdata/telegraf#4393]
Regression on the master branch. Telegraf used to run cleanly with the -race flag turned on. At or before commit a0ece79 race conditions are detected in the Accumulator/models code. See log below. This can be reproduced MacOS with just the standard system agent and a file output. It appears there's a race between some initialization code and the implementation of Gather().

### Relevant telegraf.conf:

System agent and file output enabled.

### System info:

MacOS: 10.12.6
Telegraf commit: a0ece79

### Steps to reproduce:
1. Pull and compile commit a0ece79
2. Run with "go run -race"

### Expected behavior:
No race conditions

### Actual behavior:
Race  conditions detected.
```
==================
WARNING: DATA RACE
Read at 0x00c4200cb720 by goroutine 28:
  github.com/influxdata/telegraf/internal/models.(*RunningInput).MakeMetric()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:71 +0x10b
  github.com/influxdata/telegraf/agent.(*accumulator).AddGauge()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/accumulator.go:63 +0x157
  github.com/influxdata/telegraf/plugins/inputs/system.(*DiskStats).Gather()
      /Users/prydin/go/src/github.com/influxdata/telegraf/plugins/inputs/system/disk.go:76 +0xb38
  github.com/influxdata/telegraf/agent.gatherWithTimeout.func1()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:153 +0x66

Previous write at 0x00c4200cb720 by main goroutine:
  github.com/influxdata/telegraf/agent.(*Agent).Run()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:101 +0x8cb
  main.reloadLoop()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:193 +0xa0
  main.main()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:364 +0xa50

Goroutine 28 (running) created at:
  github.com/influxdata/telegraf/agent.gatherWithTimeout()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:152 +0xe4
  github.com/influxdata/telegraf/agent.(*Agent).gatherer()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:124 +0x3c6
  github.com/influxdata/telegraf/agent.(*Agent).Run.func3()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:400 +0x8a
==================
WARNING: DATA RACE
Read at 0x00c4200cbb60 by goroutine 44:
  github.com/influxdata/telegraf/internal/models.(*RunningInput).MakeMetric()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:71 +0x10b
  github.com/influxdata/telegraf/agent.(*accumulator).AddGauge()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/accumulator.go:63 +0x157
  github.com/influxdata/telegraf/plugins/inputs/system.(*MemStats).Gather()
      /Users/prydin/go/src/github.com/influxdata/telegraf/plugins/inputs/system/memory.go:40 +0xa81
  github.com/influxdata/telegraf/agent.gatherWithTimeout.func1()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:153 +0x66

Previous write at 0x00c4200cbb60 by main goroutine:
  github.com/influxdata/telegraf/agent.(*Agent).Run()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:101 +0x8cb
  main.reloadLoop()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:193 +0xa0
  main.main()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:364 +0xa50

Goroutine 44 (running) created at:
  github.com/influxdata/telegraf/agent.gatherWithTimeout()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:152 +0xe4
  github.com/influxdata/telegraf/agent.(*Agent).gatherer()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:124 +0x3c6
  github.com/influxdata/telegraf/agent.(*Agent).Run.func3()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:400 +0x8a
==================
WARNING: DATA RACE
Read at 0x00c4200cbae0 by goroutine 46:
  github.com/influxdata/telegraf/internal/models.(*RunningInput).MakeMetric()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:71 +0x10b
  github.com/influxdata/telegraf/agent.(*accumulator).AddCounter()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/accumulator.go:74 +0x157
  github.com/influxdata/telegraf/plugins/inputs/system.(*DiskIO).Gather()
      /Users/prydin/go/src/github.com/influxdata/telegraf/plugins/inputs/system/diskio.go:131 +0x93a
  github.com/influxdata/telegraf/agent.gatherWithTimeout.func1()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:153 +0x66

Previous write at 0x00c4200cbae0 by main goroutine:
  github.com/influxdata/telegraf/agent.(*Agent).Run()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:101 +0x8cb
  main.reloadLoop()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:193 +0xa0
  main.main()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:364 +0xa50

Goroutine 46 (running) created at:
  github.com/influxdata/telegraf/agent.gatherWithTimeout()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:152 +0xe4
  github.com/influxdata/telegraf/agent.(*Agent).gatherer()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:124 +0x3c6
  github.com/influxdata/telegraf/agent.(*Agent).Run.func3()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:400 +0x8a
==================
WARNING: DATA RACE
Read at 0x00c4200cb820 by goroutine 32:
  github.com/influxdata/telegraf/internal/models.(*RunningInput).MakeMetric()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:71 +0x10b
  github.com/influxdata/telegraf/agent.(*accumulator).AddGauge()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/accumulator.go:63 +0x157
  github.com/influxdata/telegraf/plugins/inputs/system.(*SwapStats).Gather()
      /Users/prydin/go/src/github.com/influxdata/telegraf/plugins/inputs/system/swap.go:36 +0x56f
  github.com/influxdata/telegraf/agent.gatherWithTimeout.func1()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:153 +0x66

Previous write at 0x00c4200cb820 by main goroutine:
  github.com/influxdata/telegraf/agent.(*Agent).Run()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:101 +0x8cb
  main.reloadLoop()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:193 +0xa0
  main.main()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:364 +0xa50

Goroutine 32 (running) created at:
  github.com/influxdata/telegraf/agent.gatherWithTimeout()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:152 +0xe4
  github.com/influxdata/telegraf/agent.(*Agent).gatherer()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:124 +0x3c6
  github.com/influxdata/telegraf/agent.(*Agent).Run.func3()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:400 +0x8a
==================
WARNING: DATA RACE
Read at 0x00c4200cb8a0 by goroutine 39:
  github.com/influxdata/telegraf/internal/models.(*RunningInput).MakeMetric()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:71 +0x10b
  github.com/influxdata/telegraf/agent.(*accumulator).AddGauge()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/accumulator.go:63 +0x157
  github.com/influxdata/telegraf/plugins/inputs/system.(*SystemStats).Gather()
      /Users/prydin/go/src/github.com/influxdata/telegraf/plugins/inputs/system/system.go:48 +0x457
  github.com/influxdata/telegraf/agent.gatherWithTimeout.func1()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:153 +0x66

Previous write at 0x00c4200cb8a0 by main goroutine:
  github.com/influxdata/telegraf/agent.(*Agent).Run()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:101 +0x8cb
  main.reloadLoop()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:193 +0xa0
  main.main()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:364 +0xa50

Goroutine 39 (running) created at:
  github.com/influxdata/telegraf/agent.gatherWithTimeout()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:152 +0xe4
  github.com/influxdata/telegraf/agent.(*Agent).gatherer()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:124 +0x3c6
  github.com/influxdata/telegraf/agent.(*Agent).Run.func3()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:400 +0x8a
==================
WARNING: DATA RACE
Read at 0x00c4200cbbe0 by goroutine 48:
  github.com/influxdata/telegraf/internal/models.(*RunningInput).MakeMetric()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:71 +0x10b
  github.com/influxdata/telegraf/agent.(*accumulator).AddGauge()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/accumulator.go:63 +0x157
  github.com/influxdata/telegraf/plugins/inputs/system.(*Processes).Gather()
      /Users/prydin/go/src/github.com/influxdata/telegraf/plugins/inputs/system/processes.go:61 +0xd4
  github.com/influxdata/telegraf/agent.gatherWithTimeout.func1()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:153 +0x66

Previous write at 0x00c4200cbbe0 by main goroutine:
  github.com/influxdata/telegraf/agent.(*Agent).Run()
      /Users/prydin/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:101 +0x8cb
  main.reloadLoop()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:193 +0xa0
  main.main()
      /Users/prydin/go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:364 +0xa50

Goroutine 48 (running) created at:
  github.com/influxdata/telegraf/agent.gatherWithTimeout()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:152 +0xe4
  github.com/influxdata/telegraf/agent.(*Agent).gatherer()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:124 +0x3c6
  github.com/influxdata/telegraf/agent.(*Agent).Run.func3()
      /Users/prydin/go/src/github.com/influxdata/telegraf/agent/agent.go:400 +0x8a
==================
```
### Additional info:

N/A

**Comment by danielnelson** [comment]
Thanks for the report, I believe this stems back to https://github.com/influxdata/telegraf/pull/4230.  Can you test with the fix in #4395?

**Comment by prydin** [comment]
Yup. #4395 fixed the problem for me. Thanks!
